### PR TITLE
Describe a self-service policy investigation

### PR DESCRIPTION
**Pull request by smarterclayton**
Logical clusters offer opportunities for efficient and strong
multi-tenancy. Explore some of the possible use cases that would
benefit large deployers of Kubernetes applications by standardizing
self-service and exposing new policy constructs that allow better
subdivision.

Addresses #138

### CODE REVIEW

File: docs/investigations/self-service-policy.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+## Areas of investigation
+
+* Using logical clusters as a mechanism for tenancy, but having a backing implementation that can change
+  * I.e. materialize logical clusters as an API resource in a separate logical cluster
+  * Or implementing logical clusters outside the system and having the kcp server implementation be a shim
+  * Formal "policy module" implementations that can be plugged into a minimal API server while using logical cluster impl
+* Catalog the set of tenancy constructs in use in Kube
+  * Draw heavily on sig-multitenancy explorations - work done by cluster api nested, namespace tenancy, hierarchal namespace designs
```
Comment (vincent-pli) [inline docs/investigations/self-service-policy.md:33]: Could you add the link to these design, for example:
[cluster api nested](https://github.com/kubernetes-sigs/cluster-api-provider-nested/tree/main/virtualcluster)

File: docs/investigations/self-service-policy.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+## Problem:
+
+A key component of large Kubernetes clusters is shared use, where the usage pattern might vary from externally controlled (via gitops / existing operational tools) to a permissive self-service model.  The most common partitioning model in Kubernetes is namespace, and the second most common model is cluster.  Self-service is currently limited by the set of resources that are namespace scoped for the former, and by the need to parameterize and configure multiple clusters consistently for the later. Cluster partitioning can uniquely offer distinct sets of APIs to consumers. Namespace partitioning is cheap up until the scale limits of the cluster (~10k namespaces), while cluster partitioning usually has a fixed cost per cluster in operational and resource usage, as well as lower total utilization.  Once a deployment reaches the scale limit of a single cluster, operators often need to redefine their policies and tools to work in a multi-cluster environment. Many large deployers create their own systems for managing self-service policy above their clusters and leverage individual subsystems within Kubernetes to accomplish those goals.
```
Comment (imjasonh) [inline docs/investigations/self-service-policy.md:9]:
```suggestion
A key component of large Kubernetes clusters is shared use, where the usage pattern might vary from externally controlled (via gitops / existing operational tools) to a permissive self-service model.  The most common partitioning model in Kubernetes is namespace, and the second most common model is cluster.

Self-service is currently limited by the set of resources that are namespace scoped for the former, and by the need to parameterize and configure multiple clusters consistently for the latter. Cluster partitioning can uniquely offer distinct sets of APIs to consumers. Namespace partitioning is cheap up until the scale limits of the cluster (~10k namespaces), while cluster partitioning usually has a fixed cost per cluster in operational and resource usage, as well as lower total utilization.

Once a deployment reaches the scale limit of a single cluster, operators often need to redefine their policies and tools to work in a multi-cluster environment. Many large deployers create their own systems for managing self-service policy above their clusters and leverage individual subsystems within Kubernetes to accomplish those goals.
```

File: docs/investigations/self-service-policy.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+## Approach:
+
+The logical cluster concept offers an opportunity to allow self-service at a cluster scope, with the effective cost of the namespace partitioning scheme. In addition, the separation of workload at control plane (kcp) and data plane (physical cluster) via transparent multi-cluster or similar schemes allows strong policy control of what configuration is allowed (reject early), restriction of the supported API surface area for workload APIs (limit / control certain fields like pod security), and limits the access of individual users to the underlying infra (much like clusters limit access to nodes).
```
Comment (imjasonh) [inline docs/investigations/self-service-policy.md:14]:
```suggestion
The logical cluster concept offers an opportunity to allow self-service at a cluster scope, with the effective cost of the namespace partitioning scheme. In addition, the separation of workload at control plane (kcp) and data plane (physical cluster) via [transparent multi-cluster](./transparent-multi-cluster.md) or similar schemes allows strong policy control of what configuration is allowed (reject early), restriction of the supported API surface area for workload APIs (limit / control certain fields like pod security), and limits the access of individual users to the underlying infra (much like clusters limit access to nodes).
```

**Review by imjasonh — APPROVED** [review]
lgtm, just some small tweaks for readability.

Did you intend to add more content before merging? This is a fine base to start from.

File: docs/investigations/self-service-policy.md (hunk 1 above)
Comment (smarterclayton) [reply docs/investigations/self-service-policy.md:37]: Definitely.

**Pull request merged**